### ISSUE TITLE
encoder decoder

### ISSUE THREAD

**Issue by DNikhilesh122** [DNikhilesh122/seq-to-seq-chatbot#2]
import torch
import torch.nn as nn

class Encoder(nn.Module):
    def __init__(self, vocab_size, emb=256, hidden=512):
        super().__init__()
        self.embedding = nn.Embedding(vocab_size, emb)
        self.lstm = nn.LSTM(emb, hidden, batch_first=True)

    def forward(self, x):
        emb = self.embedding(x)
        outputs, (hidden, cell) = self.lstm(emb)
        return outputs, hidden, cell

class Attention(nn.Module):
    def __init__(self, hidden):
        super().__init__()
        self.W = nn.Linear(hidden * 2, hidden)
        self.V = nn.Linear(hidden, 1)

    def forward(self, hidden, encoder_outputs):
        seq_len = encoder_outputs.shape[1]
        hidden = hidden.repeat(seq_len, 1, 1).permute(1, 0, 2)
        energy = torch.tanh(self.W(torch.cat((hidden, encoder_outputs), dim=2)))
        attention = self.V(energy).squeeze(2)
        return torch.softmax(attention, dim=1)

class Decoder(nn.Module):
    def __init__(self, vocab_size, hidden=512, emb=256):
        super().__init__()
        self.embedding = nn.Embedding(vocab_size, emb)
        self.attention = Attention(hidden)
        self.lstm = nn.LSTM(hidden + emb, hidden, batch_first=True)
        self.fc = nn.Linear(hidden, vocab_size)

    def forward(self, x, hidden, cell, encoder_outputs):
        x = x.unsqueeze(1)
        emb = self.embedding(x)

        attn_weights = self.attention(hidden[-1], encoder_outputs)
        context = torch.bmm(attn_weights.unsqueeze(1), encoder_outputs)

        lstm_input = torch.cat((emb, context), dim=2)
        output, (hidden, cell) = self.lstm(lstm_input, (hidden, cell))

        prediction = self.fc(output.squeeze(1))
        return prediction, hidden, cell
